### ISSUE TITLE
Port to AngularJS

### ISSUE THREAD

**Issue by raphink** [camptocamp/puppet-catalog-diff-viewer#11]
Expected gain:
- Simpler code, using templates
- Maybe a perf gain (from templates)
- Routing, allowing for shareable URLs, such as `/#/my_report/changes/foo.example.com/diff/dfere4358df`
# Routes

Here are the routes that need to be implemented:
- `/:report`: load a report
- `/:report/failed`: display failed nodes
- `/:report/failed/:node`: display one failed node
- `/:report/changes`: display nodes with changes
- `/:report/changes/:node`: display one node with changes
- `/:report/changes/:node/diff`: jump to diff panel
- `/:report/changes/:node/diff/:resource`: jump to a resource in the diff panel
- `/:report/changes/:node/in-old`: jump to in-old panel
- `/:report/changes/:node/in-old/:resource`: jump to a resource in the in-old panel
- `/:report/changes/:node/in-new`: jump to in-new panel
- `/:report/changes/:node/in-new/:resource`: jump to a resource in the in-new panel
- `/starred` (see https://github.com/camptocamp/puppet-catalog-diff-viewer/issues/14)
- `/acked` (see https://github.com/camptocamp/puppet-catalog-diff-viewer/issues/14)
## Resource IDs

Resources will need to be identified uniquely within a section (diff, in-old, in-new). There resource full name (`type[title]`) is sure to be unique, but not very easy to pass in a hash (it can contain spaces and special characters), so we might want to hash it:
- URL encoding might do it, but it won't look very good in the URL
- Base64 it:

``` javascript
> btoa("filesystem[/dev/mapper/vg0-pg_backups]")
"ZmlsZXN5c3RlbVsvZGV2L21hcHBlci92ZzAtcGdfYmFja3Vwc10="
```
- using a hashing function, such as [hashCode()](http://stackoverflow.com/questions/7616461/generate-a-hash-from-string-in-javascript-jquery) or [murmur](https://github.com/karanlyons/murmurHash3.js) (or https://github.com/garycourt/murmurhash-js, faster but only supports 32 bit hashes), might be nicer if it returns unique IDs. Several problems with `hashCode()`:
  - it can return negative numbers => apply `Math.abs()` to it. This doubles the chances of collisions, but they are still quite unlikely
  - it returns only numbers => apply `.toString(36)` to it. 

In short, for example:

``` javascript
> murmurhash3_32_gc("file[/etc/bind/dynamic/compute.camptocamp.com.conf]").toString(36)
"xagm9b"
```

If we have a nice, safe hashing function, we might actually want to use it internally for resource referrers, instead of trying to escape the full resource names.
# Templates

Several templates are required:
- `chart.html`: report chart template, generating the d3js markup
- `nodes_list.html`: nodes list template, with links to node routes
- `node.html`: node details template
# Caching

As the global variables will be reimplemented in the `$scope`, it would be good to consider caching all the reports in there, and having a button to refresh reports. This way, reports and acks/stars will be kept in memory for all reports during the time of the session.
